apps/src/lib/kits/maker/ui/BoardSetupCheck.jsx

```diff
@@ -0,0 +1,335 @@
+/** @file Maker Board setup checker */
+import React, {Component} from 'react';
+import CircuitPlaygroundBoard from '../CircuitPlaygroundBoard';
+import {ensureAppInstalled, findPortWithViableDevice} from '../portScanning';
+import SetupStep, {
+  HIDDEN,
+  WAITING,
+  ATTEMPTING,
+  SUCCEEDED,
+  FAILED,
+  CELEBRATING,
+} from './SetupStep';
+
+const STATUS_IS_CHROME = 'statusIsChrome';
+const STATUS_APP_INSTALLED = 'statusAppInstalled';
+const STATUS_WINDOWS_DRIVERS = 'statusWindowsDrivers';
+const STATUS_BOARD_PLUG = 'statusBoardPlug';
+const STATUS_BOARD_CONNECT = 'statusBoardConnect';
+const STATUS_BOARD_COMPONENTS = 'statusBoardComponents';
+
+const initialState = {
+  isDetecting: false,
+  caughtError: null,
+  [STATUS_IS_CHROME]: WAITING,
+  [STATUS_APP_INSTALLED]: WAITING,
+  [STATUS_WINDOWS_DRIVERS]: WAITING,
+  [STATUS_BOARD_PLUG]: WAITING,
+  [STATUS_BOARD_CONNECT]: WAITING,
+  [STATUS_BOARD_COMPONENTS]: WAITING,
+};
+
+export default class BoardSetupCheck extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {...initialState};
+  }
+
+  hide(selector) {
+    this.setState({[selector]: HIDDEN});
+  }
+
+  fail(selector) {
+    this.setState({[selector]: FAILED});
+  }
+
+  spin(selector) {
+    this.setState({[selector]: ATTEMPTING});
+  }
+
+  succeed(selector) {
+    this.setState({[selector]: SUCCEEDED});
+  }
+
+  thumb(selector) {
+    this.setState({[selector]: CELEBRATING});
+  }
+
+  getSurveyURL() {
+    const baseFormURL = 'https://docs.google.com/forms/d/e/1FAIpQLSe4NB7weq20sydf4kKn3QzIIn1O91hfPNU0U6b2xc1W6w44eQ/viewform';
+    const userAgentFieldFill = `entry.1520933088=${encodeURIComponent(navigator.userAgent)}`;
+    const prettifiedCurrentStates = JSON.stringify(this.state, null, 2);
+    const setupStatesFieldFill = `entry.804069894=${encodeURIComponent(prettifiedCurrentStates)}`;
+    return `${baseFormURL}?${userAgentFieldFill}&${setupStatesFieldFill}`;
+  }
+
+  detect() {
```

Comment:
I feel like all these detecter functions should live outside of the component. It seems like there is a lot of non-display logic in them and it would be easier to unit test separately. What do you think?